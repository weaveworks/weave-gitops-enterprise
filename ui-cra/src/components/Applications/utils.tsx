import { useListSources } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import { useMemo } from 'react';
import { GitopsCluster } from '../../cluster-services/cluster_services.pb';
import { request } from '../../utils/request';

const toCluster = (clusterName: string): GitopsCluster => {
  const [firstBit, secondBit] = clusterName.split('/');
  const [namespace, name, controlPlane] = secondBit
    ? [firstBit, secondBit, false]
    : ['', firstBit, true];
  return {
    name,
    namespace,
    controlPlane,
  };
};
export const useClustersWithSources = (
  allowSelectCluster: boolean,
): GitopsCluster[] | undefined => {
  const { data } = useListSources();
  const clusters = useMemo(() => {
    return _.uniq(data?.result?.map(s => s.clusterName))
      .sort()
      .map(toCluster);
  }, [data]);
  return allowSelectCluster ? clusters : undefined;
};

export const useIsClusterWithSources = (clusterName: string): boolean => {
  const clusters = useClustersWithSources(true);
  return clusters?.some((c: GitopsCluster) => c.name === clusterName) || false;
};

export const AddApplicationRequest = ({ ...data }, token: string | null) => {
  return request('POST', `/v1/enterprise/automations`, {
    body: JSON.stringify(data),
    headers: new Headers({ 'Git-Provider-Token': `token ${token}` }),
  });
};

export const renderKustomization = (data: any) => {
  return request('POST', `/v1/enterprise/automations/render`, {
    body: JSON.stringify(data),
  });
};
