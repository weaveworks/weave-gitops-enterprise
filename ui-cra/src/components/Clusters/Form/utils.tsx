import { GitopsClusterEnriched } from '../../../types/custom';

export const maybeParseJSON = (data: string) => {
  try {
    return JSON.parse(data);
  } catch (e) {
    // FIXME: show a warning to a user or something
    return undefined;
  }
};

export const getCreateRequestAnnotation = (cluster: GitopsClusterEnriched) => {
  return (
    cluster?.annotations &&
    maybeParseJSON(cluster?.annotations['templates.weave.works/create-request'])
  );
};
