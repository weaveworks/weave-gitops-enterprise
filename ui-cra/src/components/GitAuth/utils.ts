import { PageRoute } from '@weaveworks/weave-gitops/ui/lib/types';
import { GitProvider } from '../../api/gitauth/gitauth.pb';

const tokenKey = (providerName: GitProvider) =>
  `gitProviderToken_${providerName}`;

export function storeProviderToken(providerName: GitProvider, token: string) {
  if (!window.localStorage) {
    console.warn('no local storage found');
    return;
  }
  localStorage.setItem(tokenKey(providerName), token);
}

export function getProviderToken(providerName: GitProvider): string | null {
  if (!window.localStorage) {
    console.warn('no local storage found');
    return null;
  }
  return localStorage.getItem(tokenKey(providerName));
}

const CALLBACK_STATE_KEY = 'oauth_callback_state';

export type CallbackSessionState = {
  page: PageRoute | string;
  state: any;
} | null;

export function storeCallbackState(value: CallbackSessionState) {
  if (!window.sessionStorage) {
    console.warn('no session storage found');
    return;
  }

  if (!value) {
    return null;
  }

  return sessionStorage.setItem(CALLBACK_STATE_KEY, JSON.stringify(value));
}

export function getCallbackState(): CallbackSessionState {
  const state = sessionStorage.getItem(CALLBACK_STATE_KEY);

  if (!state) {
    return null;
  }

  try {
    return JSON.parse(state);
  } catch (e) {
    return null;
  }
}

export function clearCallbackState() {
  return sessionStorage.removeItem(CALLBACK_STATE_KEY);
}

export enum GrpcErrorCodes {
  Unknown = 2,
  Unauthenticated = 16,
  NotFound = 5,
}

export const navigate = (url: string) => {
  if (process.env.NODE_ENV === 'test') {
    return;
  }
  window.location.href = url;
};
