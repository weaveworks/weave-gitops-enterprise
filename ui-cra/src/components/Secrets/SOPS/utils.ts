import { GitRepository } from '@weaveworks/weave-gitops';
export interface SOPS {
  clusterName: string;
  secretName: string;
  secretNamespace: string;
  encryptionType: string;
  kustomization: string;
  data: { id: number; key: string; value: string }[];
  repo: string | null | GitRepository;
  provider: string;
  branchName: string;
  pullRequestTitle: string;
  commitMessage: string;
  pullRequestDescription: string;
}
export function getInitialData(
  callbackState: { state: { formData: SOPS } } | null,
  random: string,
) {
  let defaultFormData = {
    repo: null,
    provider: '',
    branchName: `add-SOPS-secret-branch-${random}`,
    pullRequestTitle: 'Add SOPS Secret',
    commitMessage: 'Add SOPS Secret',
    pullRequestDescription: 'This PR adds a new SOPS Secret',
    clusterName: '',
    secretName: '',
    secretNamespace: '',
    encryptionType: 'GPG/AGE',
    kustomization: '',
    data: [{ id: 1, key: '', value: '' }],
  };

  const initialFormData = {
    ...defaultFormData,
    ...callbackState?.state?.formData,
  };

  return { initialFormData };
}
export const convertToObject = (
  arr: {
    key: string;
    value: string;
  }[],
) => {
  const obj: { [key: string]: string } = {};
  arr.forEach(o => {
    obj[o.key] = o.value;
  });
  return obj;
};
export function scrollToAlertSection() {
  const element = document.getElementsByClassName('MuiAlert-root')[0];
  element?.scrollIntoView({ behavior: 'smooth' });
}

export const handleError = (err: any, setNotifications: any) => {
  if (err.code === 401) {
    const { pathname, search } = window.location;
    const redirectUrl = encodeURIComponent(`${pathname}${search}`);
    const url = redirectUrl
      ? `/sign_in?redirect=${redirectUrl}`
      : `/sign_in?redirect=/`;
    window.location.href = url;
  }
  setNotifications([
    {
      message: { text: err.message },
      severity: 'error',
      display: 'top',
    },
  ]);
  scrollToAlertSection();
};

export const getFormattedPayload = (formData: SOPS) => {
  const {
    clusterName,
    secretName,
    secretNamespace,
    kustomization,
    data,
  } = formData;
  const [k_name, k_namespace] = kustomization.split('/');
  const [c_namespace, c_name] = clusterName.split('/');
  const cluster =
    clusterName.split('/').length > 1
      ? {
          name: c_name,
          namespace: c_namespace,
        }
      : {
          name: c_namespace,
        };

  return {
    encryptionPayload: {
      clusterName,
      name: secretName,
      namespace: secretNamespace,
      kustomization_name: k_name,
      kustomization_namespace: k_namespace,
      data: convertToObject(data),
    },
    cluster,
  };
};
