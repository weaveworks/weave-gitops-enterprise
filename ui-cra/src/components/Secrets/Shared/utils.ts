import { GitRepository } from '@weaveworks/weave-gitops';

export interface ExternalSecret {
  secretStoreType: any;
  secretPath: string | undefined;
  secretStore: string;
  dataSecretKey: string | undefined;
  secretStoreKind: string | undefined;
  secretStoreRef: string;
  clusterName: string;
  secretName: string;
  secretNamespace: string;
  includeAllProps: boolean;
  data: { id: number; key: string; value: string }[];

  repo: string | null | GitRepository;
  provider: string;
  branchName: string;
  pullRequestTitle: string;
  commitMessage: string;
  pullRequestDescription: string;
}
export function getESInitialData(
  callbackState: { state: { formData: ExternalSecret } } | null,
  random: string,
) {
  let defaultFormData = {
    repo: null,
    provider: '',
    branchName: `add-external-secret-branch-${random}`,
    pullRequestTitle: 'Add External Secret',
    commitMessage: 'Add External Secret',
    pullRequestDescription: 'This PR adds a new External Secret',
    dataSecretKey: '',
    secretStoreKind: '',
    secretStoreRef: '',
    clusterName: '',
    secretName: '',
    secretNamespace: '',
    secretStoreType: '',
    secretStore: '',
    secretPath: '',
    includeAllProps: false,
    data: [{ id: 1, key: '', value: '' }],
  };

  const initialFormData = {
    ...defaultFormData,
    ...callbackState?.state?.formData,
  };

  return { initialFormData };
}

export interface SOPS {
  clusterName: string;
  secretName: string;
  secretNamespace: string;
  encryptionType: string;
  kustomization: string;
  data: { id: number; key: string; value: string }[];
  repo: string | null | GitRepository;
  provider: string;
  branchName: string;
  pullRequestTitle: string;
  commitMessage: string;
  pullRequestDescription: string;
}
export function getInitialData(
  callbackState: { state: { formData: SOPS } } | null,
  random: string,
) {
  let defaultFormData = {
    repo: null,
    provider: '',
    branchName: `add-SOPS-secret-branch-${random}`,
    pullRequestTitle: 'Add SOPS Secret',
    commitMessage: 'Add SOPS Secret',
    pullRequestDescription: 'This PR adds a new SOPS Secret',
    clusterName: '',
    secretName: '',
    secretNamespace: '',
    encryptionType: 'GPG/AGE',
    kustomization: '',
    data: [{ id: 1, key: '', value: '' }],
  };

  const initialFormData = {
    ...defaultFormData,
    ...callbackState?.state?.formData,
  };

  return { initialFormData };
}
export const convertToObject = (
  arr: {
    key: string;
    value: string;
  }[],
) => {
  const obj: { [key: string]: string } = {};
  arr.forEach(o => {
    obj[o.key] = o.value;
  });
  return obj;
};
export function scrollToAlertSection() {
  const element = document.getElementsByClassName('MuiAlert-root')[0];
  element?.scrollIntoView({ behavior: 'smooth' });
}

export const handleError = (err: any, setNotifications: any) => {
  if (err.code === 401) {
    const { pathname, search } = window.location;
    const redirectUrl = encodeURIComponent(`${pathname}${search}`);
    const url = redirectUrl
      ? `/sign_in?redirect=${redirectUrl}`
      : `/sign_in?redirect=/`;
    window.location.href = url;
  }
  setNotifications([
    {
      message: { text: err.message },
      severity: 'error',
      display: 'top',
    },
  ]);
  scrollToAlertSection();
};

export const getFormattedPayload = (formData: SOPS) => {
  const { clusterName, secretName, secretNamespace, kustomization, data } =
    formData;
  const [k_name, k_namespace] = kustomization.split('/');
  const [c_namespace, c_name] = clusterName.split('/');
  const cluster =
    clusterName.split('/').length > 1
      ? {
          name: c_name,
          namespace: c_namespace,
        }
      : {
          name: c_namespace,
        };

  return {
    encryptionPayload: {
      clusterName,
      name: secretName,
      namespace: secretNamespace,
      kustomization_name: k_name,
      kustomization_namespace: k_namespace,
      data: convertToObject(data),
    },
    cluster,
  };
};
export const getESFormattedPayload = (formData: ExternalSecret) => {
  const {
    clusterName,
    secretName,
    secretNamespace,
    secretStoreRef,
    secretStoreKind,
    dataSecretKey,
    secretPath,
    includeAllProps,
    data,
  } = formData;
  const [c_namespace, c_name] = clusterName.split('/');
  const cluster =
    clusterName.split('/').length > 1
      ? {
          name: c_name,
          namespace: c_namespace,
        }
      : {
          name: c_namespace,
        };

  const dataObj = includeAllProps
    ? {
        dataFrom: {
          extract: {
            key: secretPath,
          },
        },
      }
    : {
        data: data.map(d => {
          return {
            secretKey: d.value,
            remoteRef: {
              key: d.key,
              property: d.value,
            },
          };
        }),
      };

  return {
    externalSecret: {
      metadata: {
        name: secretName,
        namespace: secretNamespace,
      },
      spec: {
        refreshInterval: '1h',
        secretStoreRef: {
          name: secretStoreRef,
          kind: secretStoreKind,
        },
        target: {
          name: dataSecretKey,
        },
        ...dataObj,
      },
    },
    cluster,
    isControlPlane: c_namespace ? true : false,
  };
};
