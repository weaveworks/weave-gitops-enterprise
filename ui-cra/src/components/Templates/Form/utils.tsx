import {
  Automation,
  GitRepository,
  Source,
} from '@weaveworks/weave-gitops/ui/lib/objects';
import { Pipeline } from '../../../api/pipelines/types.pb';
import { GetTerraformObjectResponse } from '../../../api/terraform/terraform.pb';
import { GitopsClusterEnriched } from '../../../types/custom';
import { Resource } from '../Edit/EditButton';
import GitUrlParse from 'git-url-parse';
import URI from 'urijs';
import { GetConfigResponse } from '../../../cluster-services/cluster_services.pb';

const yamlConverter = require('js-yaml');

export const maybeParseJSON = (data: string) => {
  try {
    return JSON.parse(data);
  } catch (e) {
    // FIXME: show a warning to a user or something
    return undefined;
  }
};

export const getCreateRequestAnnotation = (resource: Resource) => {
  const getAnnotation = (resource: Resource) => {
    switch (resource.type) {
      case 'GitopsCluster':
        return (resource as GitopsClusterEnriched)?.annotations?.[
          'templates.weave.works/create-request'
        ];
      case 'GitRepository':
      case 'Bucket':
      case 'HelmRepository':
      case 'HelmChart':
      case 'Kustomization':
      case 'HelmRelease':
      case 'OCIRepository':
        return (resource as Automation | Source)?.obj?.metadata?.annotations?.[
          'templates.weave.works/create-request'
        ];
      case 'Terraform':
      case 'Pipeline':
        return yamlConverter.load(
          (resource as GetTerraformObjectResponse | Pipeline)?.yaml,
        )?.metadata?.annotations?.['templates.weave.works/create-request'];
      default:
        return '';
    }
  };

  return maybeParseJSON(getAnnotation(resource));
};

export function getRepositoryUrl(repo: GitRepository) {
  // the https url can be overridden with an annotation
  const httpsUrl =
    repo?.obj?.metadata?.annotations?.['weave.works/repo-https-url'];
  if (httpsUrl) {
    return httpsUrl;
  }
  let uri = URI(repo?.obj?.spec?.url);
  if (uri.hostname() === 'ssh.dev.azure.com') {
    uri = azureSshToHttps(uri.toString());
  }
  return uri.protocol('https').port('').userinfo('').toString();
}

function azureSshToHttps(sshUrl: string) {
  const parts = sshUrl.split('/');
  const organization = parts[4];
  const project = parts[5];
  const repository = parts[6];

  const httpsUrl = `https://dev.azure.com/${organization}/${project}/_git/${repository}`;

  return URI(httpsUrl);
}

export function getProvider(repo: GitRepository, config: GetConfigResponse) {
  const url = getRepositoryUrl(repo);
  const domain = URI(url).hostname();
  return config?.gitHostTypes?.[domain] || 'github';
}

export function getInitialGitRepo(
  initialUrl: string | null,
  gitRepos: GitRepository[],
) {
  // if there is a repository url in the create-pr annotation, go through the gitrepos and compare it to their links;
  // if no result, parse it and check for the protocol; if ssh, convert it to https and try again to compare it to the gitrepos links
  // createPRRepo signals that this refers to a pre-existing resource
  if (initialUrl) {
    for (var repo of gitRepos) {
      let repoUrl = repo?.obj?.spec?.url;
      if (repoUrl === initialUrl) {
        return { ...repo, createPRRepo: true };
      }
      let parsedRepolUrl = GitUrlParse(repoUrl);
      if (parsedRepolUrl?.protocol === 'ssh') {
        if (
          initialUrl === parsedRepolUrl.href.replace('ssh://git@', 'https://')
        ) {
          return { ...repo, createPRRepo: true };
        }
      }
    }
  }

  // FIXME: return getDefaultGitRepo(gitRepos, config);
  return null;
}

export function getDefaultGitRepo(
  gitRepos: GitRepository[],
  config: GetConfigResponse,
) {
  const annoRepo = gitRepos.find(
    repo =>
      repo?.obj?.metadata?.annotations?.['weave.works/repo-role'] === 'default',
  );
  if (annoRepo) {
    return annoRepo;
  }

  const mainRepo = gitRepos.find(
    repo =>
      repo.clusterName === config.managementClusterName &&
      repo?.obj?.metadata?.name === 'flux-system' &&
      repo?.obj?.metadata?.namespace === 'flux-system',
  );

  if (mainRepo) {
    return mainRepo;
  }

  return gitRepos[0];
}
