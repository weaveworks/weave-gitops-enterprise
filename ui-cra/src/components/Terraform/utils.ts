import { TerraformObject } from '../../api/terraform/types.pb';

export interface TerraformObjectNode extends TerraformObject {
  id: string;
  isCurrentNode: boolean;
  type: string;
  parentIds: string[];
}

export type TerraformNodesMap = { [key: string]: TerraformObjectNode };

export function makeObjectId(namespace?: string, name?: string) {
  return namespace + '/' + name;
}

export function getNeighborNodes(
  nodes: TerraformNodesMap,
  currentNode: TerraformObject,
): TerraformObjectNode[] {
  let dependencyNodes: TerraformObjectNode[] = [];
  if (currentNode.dependsOn) {
    dependencyNodes = currentNode.dependsOn
      .map(dependency => {
        const name = dependency.name;
        let namespace = dependency.namespace;

        if (!namespace) {
          namespace = currentNode.namespace;
        }

        return nodes[makeObjectId(namespace, name)];
      })
      .filter(n => n);
  }

  const nodesArray: TerraformObjectNode[] = Object.values(nodes);

  const dependentNodes = nodesArray.filter(node => {
    let isDependent = false;

    for (const dependency of node.dependsOn || []) {
      const name = dependency.name;
      let namespace = dependency.namespace;
      if (!namespace) {
        namespace = node.namespace;
      }

      if (name === currentNode.name && namespace === currentNode.namespace) {
        isDependent = true;
        break;
      }
    }

    return isDependent;
  });

  return dependencyNodes.concat(dependentNodes);
}

// getGraphNodes returns all nodes in the current node's dependency tree, including the current node.
export function getGraphNodes(
  nodes: TerraformNodesMap,
  object: TerraformObject,
): TerraformObjectNode[] {
  // Find node, corresponding to the object.
  const currentNode = nodes[makeObjectId(object.namespace, object.name)];

  if (!currentNode) {
    return [];
  }

  currentNode.isCurrentNode = true;

  // Find nodes in the current node's dependency tree.
  let graphNodes: TerraformObjectNode[] = [];

  const visitedNodes: { [name: string]: boolean } = {};
  visitedNodes[currentNode.id] = true;
  let nodesToExplore: TerraformObjectNode[] = [currentNode];

  while (nodesToExplore.length > 0) {
    const node = nodesToExplore.shift();

    const newNodes = getNeighborNodes(nodes, node || {}).filter(
      n => !visitedNodes[n.id],
    );

    for (const n of newNodes) {
      visitedNodes[n.id] = true;
    }

    nodesToExplore = nodesToExplore.concat(newNodes);

    graphNodes = graphNodes.concat(node || []);
  }

  if (graphNodes.length === 1) {
    return [];
  }

  return graphNodes;
}
