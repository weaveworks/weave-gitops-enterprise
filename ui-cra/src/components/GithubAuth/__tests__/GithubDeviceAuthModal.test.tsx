import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { GithubDeviceAuthModal } from '..';
import { getProviderToken, GitProvider } from '../utils';
import {
  ApplicationOverrides,
  createMockClient,
  withContext,
  withTheme,
} from '../../lib/test-utils';

describe('GithubDeviceAuthModal', () => {
  let container: any = null;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });
  describe('snapshots', () => {
    it('renders', async () => {
      const ovr = {
        GetGithubDeviceCode: () => ({ userCode: '' }),
        GetGithubAuthStatus: () => ({ accessToken: '' }),
      };
      const repoName = 'ssh://example.com/user/repo.git';

      await act(async () => {
        render(
          withTheme(
            withContext(
              <GithubDeviceAuthModal
                repoName={repoName}
                open
                onClose={() => null}
                onSuccess={() => null}
              />,
              '/',
              { applicationsClient: createMockClient(ovr) },
            ),
          ),
          container,
        );
      });

      const modal = screen.getByRole('presentation');
      expect(modal).toMatchSnapshot();
    });
  });
  it('renders the user code', async () => {
    const userCode = '123-456';
    const ovr: ApplicationOverrides = {
      GetGithubDeviceCode: () => ({ userCode }),
      GetGithubAuthStatus: () => ({ error: 'some error' }),
    };
    await act(async () => {
      render(
        withTheme(
          withContext(
            <GithubDeviceAuthModal
              repoName="ssh://example.com/user/repo.git"
              open
              onClose={() => null}
              onSuccess={() => null}
            />,
            '/',
            { applicationsClient: createMockClient(ovr) },
          ),
        ),
        container,
      );
    });

    const code = screen.getByText(userCode);
    expect(code.innerHTML).toEqual(userCode);
  });
  it('stores a token', async () => {
    const accessToken = 'sometoken';
    const ovr = {
      GetGithubDeviceCode: () => ({ userCode: '' }),
      GetGithubAuthStatus: () => ({ accessToken }),
    };
    await act(async () => {
      render(
        withTheme(
          withContext(
            <GithubDeviceAuthModal
              repoName="ssh://example.com/user/repo.git"
              open
              onClose={() => null}
              onSuccess={() => null}
            />,
            '/',
            { applicationsClient: createMockClient(ovr) },
          ),
        ),
        container,
      );
    });

    const token = getProviderToken(GitProvider.GitHub);
    expect(token).toEqual(accessToken);
  });
});
