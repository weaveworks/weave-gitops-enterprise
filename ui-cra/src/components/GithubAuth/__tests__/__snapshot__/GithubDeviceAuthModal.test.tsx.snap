// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubDeviceAuthModal snapshots renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  font-family: 'proxima-nova',Helvetica,Arial,sans-serif;
  font-size: 32px;
  font-weight: 400;
  text-transform: none;
  font-style: normal;
}

.c10 svg {
  fill: !important;
  height: 16px;
  width: 16px;
}

.c10.downward {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c10.upward {
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c10 .c5 {
  margin-left: 4px;
}

.c10 img {
  width: 16px;
}

.c7.MuiButton-root {
  height: 32px;
  font-size: 12px;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1;
  border-radius: 2px;
  font-weight: 600;
}

.c7.MuiButton-outlined {
  padding: 8px 12px;
  border-color: #d8d8d8;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  background-color: #fff;
  margin: 0 auto;
  max-width: 540px;
  padding: 16px 32px;
  -webkit-transform: translate(0,50%);
  -ms-transform: translate(0,50%);
  transform: translate(0,50%);
}

.c4 {
  padding: 8px 0;
}

.c3 .c9 {
  margin-left: 8px;
}

<div
  aria-describedby="simple-modal-description"
  aria-labelledby="simple-modal-title"
  role="presentation"
  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
>
  <div
    aria-hidden="true"
    style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: rgba(0, 0, 0, 0.5);"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="c0"
    tabindex="-1"
  >
    <div
      class="c1"
    >
      <h2
        id="simple-modal-title"
      >
        Authenticate with Github
      </h2>
      <p
        id="simple-modal-description"
      >
        Weave GitOps needs to authenticate with the Git Provider for the ssh://example.com/user/repo.git repo
      </p>
    </div>
    <div>
      <p>
        Paste this code into the Github Device Activation field to grant Weave GitOps temporary access:
      </p>
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <div
            class="c2 c4"
          >
            <span
              class="c5 c6"
            />
          </div>
          <div
            class="c2 c4"
          >
            <a
              target="_blank"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined c7 MuiButton-outlinedPrimary MuiButton-disableElevation"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <span
                    class="MuiButton-startIcon MuiButton-iconSizeMedium"
                  >
                    <div
                      class="c8 c9 c10"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                        />
                      </svg>
                    </div>
                  </span>
                  Authorize Github Access
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </a>
          </div>
          <div
            class="c2 c4"
          />
        </div>
      </div>
    </div>
    <div
      class="c11"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text c7 MuiButton-colorInherit MuiButton-disableElevation"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Close
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
